api/src/main/java/org/pmiops/workbench/interceptors/AuthInterceptor.java

```diff
@@ -89,4 +120,73 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
+   * Checks for an @AuthorityRequired annotation on the controller's handler method.
+   *
+   * There is a hierarchy of Swagger-generated interfaces/wrappers around our controllers:
+   *     FooApi (interface generated by Swagger)
+   *     FooApiController (generated by Swagger, handed to AuthInterceptor)
+   *       private FooApiDelegate delegate;
+   *     FooApiDelegate (interface generated by Swagger)
+   *     FooController implements FooApiDelegate (we implement this)
+   * We can only annotate FooController methods, but are given a FooApiController, so we use
+   * reflection to hack our way to FooController's method.
+   *
+   * @param method The ApiController (Swagger-generated) method which calls our annotated delegate.
+   */
+  private boolean hasRequiredAuthority(Method apiControllerMethod, Userinfoplus userInfo) {
+    // There's no concise way to find out what class implements the delegate interface, so instead
+    // depend on naming conventions. Essentially, this removes "Api" from the class name.
+    Pattern apiControllerPattern = Pattern.compile("(.*\\.[^.]+)Api(Controller)");
+    String apiControllerName = apiControllerMethod.getDeclaringClass().getName();
+    String controllerName = apiControllerPattern.matcher(apiControllerName).replaceAll("$1$2");
+    Class controllerClass;
+    try {
+      controllerClass = Class.forName(controllerName);
+    } catch (ClassNotFoundException e) {
+      log.log(
+          Level.SEVERE,
+          "Missing {0} by name derived from {1}. @AuthorityRequired will not be evaluted.",
+          new Object[] {controllerName, apiControllerName});
+      return true;  // Default to allowing access in case of error.
+    }
+
+    Method controllerMethod;
+    try {
+      controllerMethod = controllerClass.getMethod(
+          apiControllerMethod.getName(), apiControllerMethod.getParameterTypes());
+    } catch (NoSuchMethodException e) {
+      throw new RuntimeException(e);
+    }
+    String controllerMethodName =
+        controllerMethod.getDeclaringClass().getName() + "." + controllerMethod.getName();
+
+    AuthorityRequired req = controllerMethod.getAnnotation(AuthorityRequired.class);
+
+    if (req != null) {
+      Collection<Authority> granted = getGrantedAuthorities(userInfo.getEmail());
+      if (granted.containsAll(Arrays.asList(req.value()))) {
+        return true;
+      } else {
+        log.log(
+            Level.INFO,
+            "{0} required authorities {1} but user had only {2}.",
+            new Object[] {
+                controllerMethodName,
+                Arrays.toString(req.value()),
+                Arrays.toString(granted.toArray())});
+        return false;
+      }
+    }
+    return true;  // No @AuthorityRequired annotation found at runtime, default to allowed.
+  }
+
+  /**
+   * Returns a User's permissions as stored in the db; or an empty list if no user is found.
+   */
+  private Collection<Authority> getGrantedAuthorities(String email) {
```

Comment:
`hasRequiredAuthority` is running as part of this same interceptor, so I can't inject results of this interceptor for it, right? But I do dislike that double fetch [though as suggested above I don't think it's a practical concern at 1ms].

I could split this out as a separate interceptor, though then we'd end up with a different unnecessary step (publishing the security context right before denying access).

Reply:
I think this will work -- Provider<User> doesn't rely on the interceptor being done, just on the security context holder having been initialized (which you can do earlier in the interceptor.)

Latency to Cloud SQL will definitely be more than 1 ms. It won't be that slow, either, but better to avoid looking up the same data twice on every request.

Reply:
Ah, I follow now. Switched. That's a cleaner separation of concerns too.